### ISSUE TITLE
Fix the Wazuh agent compilation for Alpine Linux

### ISSUE THREAD

**Issue by marcossv9** [wazuh/wazuh#4592]
| 3.11 | Agent | Packages/Sources | Alpine Linux |

Is there a Wazuh agent package for Alpine Linux? Can I build it from source code?
It would be nice to work on it.

Thanks!

**Comment by jesusjimsa** [comment]
Hello, @marcossv9.

You can try this to install via sources.  

Installing dependencies:

```
apk update
apk add make
apk add gcc
apk add python-dev
apk add automake
apk add autoconf 
apk add libtool
apk add curl
apk add openssl
apk add openssl-dev
apk add libgcc
apk add --update alpine-sdk

# Possibly not needed
apk --no-cache add ca-certificates wget
wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-2.28-r0.apk
apk add glibc-2.28-r0.apk
```

Install linux-headers and make this symbolic link (needed later):

```
apk add linux-headers
ln -s /usr/include/linux/a.out.h /usr/include/a.out.h
```

Download Wazuh 3.8.2:

```
curl -Ls https://github.com/wazuh/wazuh/archive/v3.8.2.tar.gz | tar zx
```

Build external dependencies:
```
cd wazuh/src
make deps
```

Trick for OpenSSL regarding of POSIX:

Edit file `wazuh/src/external/openssl/crypto/async/arch/async_posix.h`, replace this:

```
# if _POSIX_VERSION >= 200112L
```

with:

```
# if _POSIX_VERSION >= 200112L \
     && (_POSIX_VERSION < 200809L || defined(__GLIBC__))
```

Now clean just the OpenSSL:

```
cd wazuh/external/openssl
make clean
```

Create group ossec and their users (ossec, ossecr and ossecm):

```
addgroup -S ossec 
adduser -S ossec -G ossec
adduser -S ossecr -G ossec
adduser -S ossecm -G ossec
```

Compile the Wazuh agent:

```
cd wazuh/src
make TARGET=server
```

Once finished, install it:

```
cd wazuh
./install.sh
```

**Comment by marcossv9** [comment]
Hello @jesusjimsa ! Thanks for the fast response.

In step "Compile the Wazuh agent" I think it should be:

`make TARGET=agent`

instead of:

`make TARGET=server`

Because I'm trying to install the wazuh agent on a container running alpine linux to be able to be monitored with a wazuh manager already running..

By the way, I executed your steps but have encountered an error and the agent never get discovered by the manager running wazuh 3.11:

/wazuh-3.8.2 # /var/ossec/bin/ossec-control start
2020/02/12 14:14:10 ossec-agentd: CRITICAL: (1751): File client.keys not found or empty.
ossec-agentd: Configuration error. Exiting


Also I still think that it would be great to have an alpine linux package for wazuh version 3.11 to get the agent installed without so many manual steps.. Alpine Linux is a widely used OS to run containers so I think it's worth it to develop.

Thanks again!

**Comment by jesusjimsa** [comment]
Yeah, you are right, I wrote the wrong one, I have edited it now to say `agent`.  

About the start of the agent, have you registered it with the manager? You can find a guide to do so here: https://documentation.wazuh.com/3.11/user-manual/registering/cli/using-command-line-linux.html#agent  

Regarding the Alpine Linux Wazuh package, it is in our roadmap and we will make it as soon as we can. However, I you want, feel free to open a Pull Request in the [wazuh-packages](https://github.com/wazuh/wazuh-packages) repository.

**Comment by marcossv9** [comment]
Hi @jesusjimsa !

Thanks for the additional information.
I managed to solve the installation issues and now it's working.

In addition to that I got some errors trying to do the steps in a Dockerfile in order to automate the process but some of them didn't work so I had to do it manually from inside the container. For example:

```
 => CACHED [10/12] RUN curl -Ls https://github.com/wazuh/wazuh/archive/v3.8.2.tar.gz | tar zx                                             0.0s
 => CACHED [11/12] RUN cd wazuh-3.8.2/src                                                                                                 0.0s
 => ERROR [12/12] RUN make deps                                                                                                           1.7s
------
 > [12/12] RUN make deps:
#16 1.456 make: *** No rule to make target 'deps'.  Stop.
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c make deps]: runc did not terminate sucessfully
```

Sadly, we need a more automated process to work with wazuh agents on alpine containers ( something like a package) because we have many containers running multiple services that sometimes scales up and then scale down and can not be managed manually.

In the other hand we managed to get it working with debian containers and everything automated. Don't need to touch the manager on that process..

I hope you guys can find some time to work in alpine package. (I'd love to help with an alpine package creation for wazuh but I have no idea where to start and how, I only code in python and bash)

Thanks for the help!
Regards,

**Comment by jesusjimsa** [comment]
I'm glad you could make it work, although it wasn't with Alpine Linux as you intended.  

I'm closing this issue now, but feel free to re-open it again if you need some more help with anything.  

Best regards,
Jesús.

**Comment by theyoyojo** [comment]
Hi,

I too am trying to build the wazuh agent on alpine linux, however after following the above instructions on a freshly installed alpine VM, I am left with the following linking error:

```
/usr/lib/gcc/x86_64-alpine-linux-musl/9.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: missing --end-group; added as last command line option
/usr/lib/gcc/x86_64-alpine-linux-musl/9.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: external/openssl/libssl.a(ssl_lib.o): relocation R_X86_64_PC32 against symbol `ssl_undefined_function' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc/x86_64-alpine-linux-musl/9.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:998: libwazuhext.so] Error 1
make[1]: Leaving directory '/root/wazuh-3.8.2/src'
make: *** [Makefile:562: agent] Error 2
```

I would attempt to modify the appropriate CFLAGS  but I am unsure how to go about doing this with the wazuh build system.

Does anyone recognize this error?

I can provide more information if needed.

Thanks,
Joel

**Comment by jesusjimsa** [comment]
Hello, @theyoyojo.

What version of Alpine are you trying to install Wazuh on?

**Comment by theyoyojo** [comment]
Hi @jesusjimsa 

Thank you for reopening this issue.

I was able to successfully install Wazuh on Alpine (latest available version) by using the latest version of Wazuh (v3.11.3 at the time I installed it) instead of the version mentioned in this issue (v3.8.2).

I no longer have any problems.

Best,
Joel

**Comment by jesusjimsa** [comment]
I'm glad you could make it work.

**Comment by mattshort** [comment]
I've walked through these steps for version 3.12.3. I'm running node:10.16.3-alpine and getting the following error:
```
CC shared/log_builder.o
shared/log_builder.c: In function 'log_builder_init':
shared/log_builder.c:54:9: warning: implicit declaration of function 'pthread_rwlockattr_setkind_np'; did you mean 'pthread_rwlockattr_setpshared'? [-Wimplicit-function-declaration]
         pthread_rwlockattr_setkind_np(&attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         pthread_rwlockattr_setpshared
shared/log_builder.c:54:46: error: 'PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP' undeclared (first use in this function)
         pthread_rwlockattr_setkind_np(&attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP);
                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
shared/log_builder.c:54:46: note: each undeclared identifier is reported only once for each function it appears in
```
Any advice on what to try would be greatly appreciated.

**Comment by okynos** [comment]
I reopened due problems related to pthread library and comments https://github.com/wazuh/wazuh/issues/4592#issuecomment-655597594 and https://github.com/wazuh/wazuh-packages/issues/451

**Comment by romelBen** [comment]
For anyone having issues like @mattshort's. I wrote a sed command that would remove the warning messages with a `sed` command:

```
sed -i '/#ifdef __linux__/,/#endif/d' shared/log_builder.c
sed -i '/#ifdef __linux__/,/#endif/d' logcollector/logcollector.c

```

These errors appear both in /wazuh-*/src/shared/log_builder.c and /wazuh-*/src/logcollector/logcollector.c. This specifically affects Alpine Linux images so I hope this will help anyone having this issue.

**Comment by moronmon** [comment]
Hi, I'm sorry to jump in, is it possible to run the Wazuh Agent on the container?

**Comment by chemamartinez** [comment]
Hi @marcossv9,

Closing this issue since we already fixed the build for Alpine Linux at https://github.com/wazuh/wazuh/pull/14726, which will be released in Wazuh v4.4.0.

For package support, it will be added at https://github.com/wazuh/wazuh-packages/pull/1715.

Thanks for your feedback.
